### ISSUE TITLE
Offset address does not seem to work correctly above level 0

### ISSUE THREAD

**Issue by nmzein** [AzHicham/openslide-rs#134]
I have been going crazy over an attempt to read level 1 of an image in chunks of a fixed size. Chunking level 0 has worked flawlessly at this point so I'm pretty sure its not a bug in my code, however, I also wrote a quick test with hard-coded values just to be sure. If you'd like to see my actual code for this please check it out [here](https://github.com/nmzein/nemato/blob/main/backend/rendering-engine/src/io.rs). Here is my attempt at reading level 1 of this [image](https://github.com/nmzein/nemato/tree/main/backend) (image.svs).

As for the issue I'm observing, its two-fold. Firstly, as the test shows, I am able to go out of bounds of the image (its dimensions are 3000x3000). This is because, secondly, the chunks overlap. So instead of each chunk containing each 1500x1500 quadrant of the image, it seems like its capturing 1500x1500 blocks that highly overlap. This is weird because this same code (linked above) worked flawlessly for level 0 where I got disjoint chunks. Also reading the entire image with 0 offset (first tile in the code below) yields the entire image. Would appreciate some insight into if I'm doing something wrong. Please note that in the GitHub repo I linked above, I have interfaced the read_image_rgb function with a read_region function of my own as I am planning to support more decoders in the future, so read_region here does not refer to the read_region in your library (the code below directly uses the library function calls).

```rust
fn test_openslide(image_path: &PathBuf) -> anyhow::Result<()> {
    let image = openslide_rs::OpenSlide::open(image_path)?;

    let tile = image.read_image_rgb(
        &openslide_rs::Region {
            size: openslide_rs::Size {
                w: 3000,
                h: 3000,
            },
            level: 1,
            address: openslide_rs::Address {
                x: 0,
                y: 0,
            },
        }
    )?;

    save_as_jpeg(&tile, "../store/test/layer_1.jpeg", 3000)?;

    let tile = image.read_image_rgb(
        &openslide_rs::Region {
            size: openslide_rs::Size {
                w: 1500,
                h: 1500,
            },
            level: 1,
            address: openslide_rs::Address {
                x: 0,
                y: 0,
            },
        },
    )?;

    save_as_jpeg(&tile, &format!("../store/test/b_tile_{}_{}_{}.jpeg", 1, 0, 0), 1500)?;

    let tile = image.read_image_rgb(
        &openslide_rs::Region {
            size: openslide_rs::Size {
                w: 1500,
                h: 1500,
            },
            level: 1,
            address: openslide_rs::Address {
                x: 0,
                y: 1500,
            },
        },
    )?;

    save_as_jpeg(&tile, &format!("../store/test/b_tile_{}_{}_{}.jpeg", 1, 0, 1500), 1500)?;

    let tile = image.read_image_rgb(
        &openslide_rs::Region {
            size: openslide_rs::Size {
                w: 1500,
                h: 1500,
            },
            level: 1,
            address: openslide_rs::Address {
                x: 1500,
                y: 0,
            },
        },
    )?;

    save_as_jpeg(&tile, &format!("../store/test/b_tile_{}_{}_{}.jpeg", 1, 1500, 0), 1500)?;

    let tile = image.read_image_rgb(
        &openslide_rs::Region {
            size: openslide_rs::Size {
                w: 1500,
                h: 1500,
            },
            level: 1,
            address: openslide_rs::Address {
                x: 1500,
                y: 1500,
            },
        },
    )?;

    save_as_jpeg(&tile, &format!("../store/test/b_tile_{}_{}_{}.jpeg", 1, 1500, 1500), 1500)?;

    let tile = image.read_image_rgb(
        &openslide_rs::Region {
            size: openslide_rs::Size {
                w: 1500,
                h: 1500,
            },
            level: 1,
            address: openslide_rs::Address {
                x: 3000,
                y: 0,
            },
        },
    )?;

    save_as_jpeg(&tile, &format!("../store/test/b_tile_{}_{}_{}.jpeg", 1, 3000, 0), 1500)?;

    Ok(())
}


fn save_as_jpeg(image: &image::RgbImage, path: &str, size: u32) -> Result<()> {
    use std::io::Write;
    //encode to jpeg
    let mut jpeg = Vec::new();

    image::codecs::jpeg::JpegEncoder::new(&mut jpeg).encode(
        image,
        size,
        size,
        image::ColorType::Rgb8,
    )?;
    
    // Open a file for writing.
    let mut file = std::fs::File::create(path)?;
    // Write the JPEG data to the file.
    file.write_all(&jpeg)?;
    Ok(())
}
```

**Comment by AzHicham** [comment]
Hello,
I’ll try to test your use case asap, in the mean time did you try with openslide python to test if you get the same results ?

**Comment by AzHicham** [comment]
I just remembered this "issue" 
FYI read_image_RGB is a wrapper around read_region which is a wrapper around the C function with the same name.
According to the documentation https://openslide.org/api/openslide_8h.html#a0eeecec682efdf2ef18c8426109acad0
You can see that the adresse (x,y) is in the level 0 reference frame. 

The behavior you are expecting is doable but you need to create a DeepZoom object then use read_image_rgb of the DeepZoom struct and not the openslide object ! It should works :)

**Comment by nmzein** [comment]
I see, thank you so much. I'll try that out as soon as I get the chance. Really appreciate your work by the way, your library's helped me out a lot.

**Comment by nmzein** [comment]
I'm a bit confused as to how to use the DeepZoomGenerator, I tried the following, but the tile I get has a length of 3.

```rust
fn test_deepzoom(image_path: &PathBuf) -> anyhow::Result<()> {
    use openslide_rs::{Address, OpenSlide, DeepZoomGenerator};

    const CHUNK_SIZE: u32 = 1024;
    const PATH: &str = "../store/test";
    
    let x = 0;
    let y = 0;
    let level = 0;

    let slide = OpenSlide::new(image_path)?;
    let image: DeepZoomGenerator<OpenSlide, _> =
        DeepZoomGenerator::new(&slide, CHUNK_SIZE, 0, false)?;

    let levels = image.level_count();
    println!("Levels: {}", levels);
    
    let tile = image.get_tile_rgb(level, Address { x, y })?;
    println!("Tile: {:?}", tile);

    save_as_jpeg(&tile, &format!("{}/tile_{}_{}_{}.jpeg", PATH, level, x, y), CHUNK_SIZE)?;

    Ok(())
}

fn save_as_jpeg(image: &image::RgbImage, path: &str, size: u32) -> anyhow::Result<()> {
    use image::codecs::jpeg::JpegEncoder;
    use std::{io::Write, fs::File};

    let mut jpeg_tile = Vec::new();

    JpegEncoder::new(&mut jpeg_tile).encode(
        image,
        size,
        size,
        image::ColorType::Rgb8,
    )?;
    
    let mut file = File::create(path)?;
    file.write_all(&jpeg_tile)?;

    Ok(())
}
```

```
Levels: 15
Tile: ImageBuffer { width: 1, height: 1, _phantom: PhantomData<image::color::Rgb<u8>>, data: [194, 140, 170] }
thread 'tokio-runtime-worker' panicked at /home/orca/.cargo/registry/src/index.crates.io-6f17d22bba15001f/image-0.24.8/src/codecs/jpeg/encoder.rs:450:9:
assertion `left == right` failed
  left: 3145728
 right: 3
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by AzHicham** [comment]
Hello, 

I didn't found time yet to debug your code sorry may be this week end. 
But note that when you use the deepzoom generator the level are inverted, and intermediate levels not existing in the original slide are created. 

For instance in the original slide you have: 
lvl0: 16x16
lvl1: 4x4

With the deepZoom generator you'll get:
lvl0: 1x1
lvl1: 2x2
...
lvl4: 16x16

You can use the DeepZoomGenerator to print all available levels. 
To be honest I was following the default behavior of the DeepZoomGenerator of the python version

**Comment by AzHicham** [comment]
Also do not trust DeepZoomGenerator ^^ 
you may ask for a certain size but receive an image with another size. This happens when you query for an none existing level in the original slide

**Comment by nmzein** [comment]
I did notice that the level count increased to 15 so my selection should still be in range. For now I went back to OpenSlide and just scaled the offset based on the level 0 dimensions and it worked so I'll be sticking with that for now.

In my opinion though the original C openslide library's choice to make the offset based on level 0 dimensions is extremely weird I don't know why they chose it to be like that. Anyways thanks for the help.
